code/modules/mob/new_player/login.dm

```diff
@@ -11,27 +18,223 @@
+			)
+	)
+
+	tgui_interact(src)
+
+/mob/new_player/tgui_interact(mob/user, datum/tgui/ui)
+	. = ..()
+
+	ui = SStgui.try_update_ui(user, src, ui)
+	if (!ui)
+		ui = new(user, src, "LobbyMenu", window = lobby_window)
+		ui.closeable = FALSE
+		ui.open(preinitialized = TRUE)
+
+/mob/new_player/ui_state(mob/user)
+	return GLOB.always_state
+
+/mob/new_player/ui_data(mob/user)
+	. = ..()
+
+	.["character_name"] = client.prefs ? client.prefs.real_name : client.key
+
+	var/postfix_text = (client.xeno_postfix) ? ("-"+client.xeno_postfix) : ""
+	var/prefix_text = (client.xeno_prefix) ? client.xeno_prefix : "XX"
+	.["xeno_prefix"] = prefix_text
+	.["xeno_postfix"] = postfix_text
+
+	.["tutorials_ready"] = SSticker?.current_state == GAME_STATE_PLAYING
+	.["round_start"] = !SSticker || !SSticker.mode || SSticker.current_state <= GAME_STATE_PREGAME
+	.["readied"] = ready
+
+	.["upp_enabled"] = GLOB.master_mode == /datum/game_mode/extended/faction_clash/cm_vs_upp::name
+	.["xenomorph_enabled"] = client.prefs && (client.prefs.get_job_priority(JOB_XENOMORPH))
+	.["predator_enabled"] = SSticker.mode?.flags_round_type & MODE_PREDATOR && SSticker.mode.check_predator_late_join(src, FALSE)
+	.["fax_responder_enabled"] = SSticker.mode?.check_fax_responder_late_join(src, FALSE)
+
+	.["lobby_author"] = SSlobby_art.author
+
+/mob/new_player/ui_assets(mob/user)
+	. = ..()
+
+	. += get_asset_datum(/datum/asset/simple/icon_states/lobby)
+	. += get_asset_datum(/datum/asset/simple/lobby_files)
+
+	if(SSlobby_art.initialized)
+		. += get_asset_datum(/datum/asset/simple/lobby_art)
+
+/mob/new_player/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+
+	switch(action)
+		if("tutorial")
+			tutorial_menu()
+			return TRUE
+
+		if("preferences")
+			// Otherwise the preview dummy will runtime
+			// because atoms aren't initialized yet
+			if(SSticker.current_state < GAME_STATE_PREGAME)
+				to_chat(src, SPAN_WARNING("Game is still starting up, please wait"))
+				return
+			if(!SSentity_manager.ready)
+				to_chat(src, SPAN_WARNING("DB is still starting up, please wait"))
+				return
+			client.prefs.ShowChoices(src)
+			return TRUE
+
+		if("playtimes")
+			if(!SSentity_manager.ready)
+				to_chat(src, SPAN_WARNING("DB is still starting up, please wait"))
+				return
+			if(client.player_data)
+				client.player_data.tgui_interact(src)
+			return TRUE
+
+		if("manifest")
+			ViewManifest()
+			return TRUE
+
+		if("hiveleaders")
+			ViewHiveLeaders()
+			return TRUE
+
+		if("late_join")
+			if(SSticker.current_state != GAME_STATE_PLAYING || !SSticker.mode)
+				to_chat(src, SPAN_WARNING("The round is either not ready, or has already finished..."))
+				return
+
+			if(SSticker.mode.flags_round_type & MODE_NO_LATEJOIN)
+				to_chat(src, SPAN_WARNING("Sorry, you cannot late join during [SSticker.mode.name]. You have to start at the beginning of the round. You may observe or try to join as an alien, if possible."))
+				return
```

Comment:
somewhat inconsistent return values here, should return FALSE if you don't intend to create an update.

same with the entire proc